File: cli-go/internal/domain/image_registry.go
package domain

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
	"time"

	"github.com/joacohbc/my-devcontainer-installer/cli-go/internal/core"
)

type ImageEntry struct {
	ProjectDir  string `json:"projectDir"`
	Workspace   string `json:"workspace"`
	Mode        string `json:"mode"`
	Image       string `json:"image"`
	Fingerprint string `json:"fingerprint,omitempty"`
	Variant     string `json:"variant,omitempty"`
	CreatedAt   string `json:"createdAt"`
	LastUpdated string `json:"lastUpdated"`
}

type registryFile struct {
	Entries []ImageEntry `json:"entries"`
}

func ImageRegistryPath() string {
	return filepath.Join(GlobalConfigDir(), "images.json")
}

func LoadRegistry() []ImageEntry {
	data, err := os.ReadFile(ImageRegistryPath())
	if err != nil {
		return nil
	}
	var rf registryFile
	if err := json.Unmarshal(data, &rf); err != nil {
		return nil
	}
	return rf.Entries
}

func SaveRegistry(entries []ImageEntry) error {
	dir := GlobalConfigDir()
	if err := os.MkdirAll(dir, 0755); err != nil {
		return err
	}
	data, err := json.MarshalIndent(registryFile{Entries: entries}, "", "  ")
	if err != nil {
		return err
	}
	data = append(data, '\n')
	return os.WriteFile(ImageRegistryPath(), data, 0644)
}

func RecordEntry(e ImageEntry) error {
	entries := LoadRegistry()
	found := false
	for i, existing := range entries {
		if existing.ProjectDir == e.ProjectDir {
			entries[i] = e
			found = true
			break
		}
	}
	if !found {
		entries = append(entries, e)
	}
	return SaveRegistry(entries)
}

func RemoveEntry(projectDir string) bool {
	entries := LoadRegistry()
	for i, e := range entries {
		if e.ProjectDir == projectDir {
			entries = append(entries[:i], entries[i+1:]...)
			_ = SaveRegistry(entries)
			return true
		}
	}
	return false
}

func FindByFingerprint(fp string) *ImageEntry {
	for _, e := range LoadRegistry() {
		if e.Fingerprint == fp {
			copy := e
			return &copy
		}
	}
	return nil
}

func ListEntries() []ImageEntry {
	return LoadRegistry()
}

var labelLineRe = regexp.MustCompile(`(?i)^LABEL\b`)

func stripLabelBlocks(dockerfile string) string {
	lines := strings.Split(dockerfile, "\n")
	out := make([]string, 0, len(lines))
	inLabel := false
	for _, line := range lines {
		trimmed := strings.TrimSpace(line)
		if !inLabel {
			if labelLineRe.MatchString(trimmed) {
				inLabel = strings.HasSuffix(trimmed, "\\")
				continue
			}
			out = append(out, line)
		} else {
			inLabel = strings.HasSuffix(trimmed, "\\")
		}
	}
	return strings.Join(out, "\n")
}

func sha256Hex(s string) string {
	h := sha256.Sum256([]byte(s))
	return hex.EncodeToString(h[:])
}

func ComputeFingerprint(dockerfileContent string, copyFileContents map[string]string, selectedModuleIDs []string) string {
	normalized := strings.TrimSpace(strings.ReplaceAll(stripLabelBlocks(dockerfileContent), "\r\n", "\n"))
	parts := []string{normalized}

	names := make([]string, 0, len(copyFileContents))
	for name := range copyFileContents {
		names = append(names, name)
	}
	sort.Strings(names)
	for _, name := range names {
		parts = append(parts, filepath.Base(name)+"\x00"+sha256Hex(copyFileContents[name]))
	}

	sortedModuleIDs := append([]string{}, selectedModuleIDs...)
	sort.Strings(sortedModuleIDs)
	parts = append(parts, "MODULES\x00"+strings.Join(sortedModuleIDs, ","))

	return sha256Hex(strings.Join(parts, "\n"))
}

func FingerprintTag(fp string) string {
	if len(fp) < 12 {
		return "devcontainer-cli/" + fp + ":latest"
	}
	return "devcontainer-cli/" + fp[:12] + ":latest"
}

// LocalImageExists reports whether a Docker image with the given reference is
// present locally, using the injected capture func.
func LocalImageExists(image string, capture CaptureFunc) bool {
	status, stdout, _ := capture([]string{"images", "-q", image})
	return status == 0 && strings.TrimSpace(stdout) != ""
}

func RecordProject(projectDir string, config *core.DevcontainerConfig, customImage string) {
	now := time.Now().UTC().Format(time.RFC3339)
	img := customImage
	if img == "" {
		img = config.Image
	}
	variant := ""
	if config.Remote != nil {
		variant = config.Remote.Variant
	}
	_ = RecordEntry(ImageEntry{
		ProjectDir:  projectDir,
		Workspace:   config.Workspace,
		Mode:        string(config.Mode),
		Image:       img,
		Variant:     variant,
		Fingerprint: config.Fingerprint,
		CreatedAt:   now,
		LastUpdated: now,
	})
}

File: cli/src/domain/image-registry.ts
import * as crypto from 'crypto';
import * as fs from 'fs';
import * as path from 'path';
import { dockerCapture } from '@/infra/docker.js';
import { globalConfigDir } from '@/domain/global-config.js';
import type { BuildMode, RemoteVariant, DevcontainerConfig } from '@/core/types.js';

export interface ImageEntry {
  projectDir: string;
  workspace: string;
  mode: BuildMode;
  image: string;
  fingerprint?: string;
  variant?: RemoteVariant;
  createdAt: string;
  lastUpdated: string;
}

interface RegistryFile {
  entries: ImageEntry[];
}

export function imageRegistryPath(): string {
  return path.join(globalConfigDir(), 'images.json');
}

export function loadRegistry(): ImageEntry[] {
  const p = imageRegistryPath();
  if (!fs.existsSync(p)) return [];
  try {
    const raw = fs.readFileSync(p, 'utf8');
    const parsed = JSON.parse(raw) as RegistryFile;
    return Array.isArray(parsed?.entries) ? parsed.entries : [];
  } catch {
    return [];
  }
}

export function saveRegistry(entries: ImageEntry[]): void {
  const dir = globalConfigDir();
  fs.mkdirSync(dir, { recursive: true });
  fs.writeFileSync(
    imageRegistryPath(),
    JSON.stringify({ entries }, null, 2) + '\n',
  );
}

export function recordEntry(e: ImageEntry): void {
  const entries = loadRegistry();
  const idx = entries.findIndex((x) => x.projectDir === e.projectDir);
  if (idx >= 0) {
    entries[idx] = { ...entries[idx], ...e, lastUpdated: e.lastUpdated };
  } else {
    entries.push(e);
  }
  saveRegistry(entries);
}

export function removeEntry(projectDir: string): boolean {
  const entries = loadRegistry();
  const idx = entries.findIndex((x) => x.projectDir === projectDir);
  if (idx < 0) return false;
  entries.splice(idx, 1);
  saveRegistry(entries);
  return true;
}

export function findByFingerprint(fingerprint: string): ImageEntry | undefined {
  return loadRegistry().find((e) => e.fingerprint === fingerprint);
}

export function listEntries(): ImageEntry[] {
  return loadRegistry();
}

// Strip LABEL blocks so label-only changes (e.g. timestamp metadata) don't
// invalidate the cache. A LABEL block is a single LABEL line or a continued
// multi-line statement (using `\` line continuations).
function stripLabelBlocks(dockerfile: string): string {
  const lines = dockerfile.split('\n');
  const out: string[] = [];
  let inLabel = false;
  for (const line of lines) {
    const trimmed = line.trim();
    if (!inLabel) {
      if (/^LABEL\b/i.test(trimmed)) {
        inLabel = trimmed.endsWith('\\');
        continue;
      }
      out.push(line);
    } else {
      inLabel = trimmed.endsWith('\\');
    }
  }
  return out.join('\n');
}

function sha256Hex(s: string | Buffer): string {
  return crypto.createHash('sha256').update(s).digest('hex');
}

export function computeFingerprint(
  dockerfileContent: string,
  copyFileContents: Record<string, string>,
  selectedModuleIds: string[] = [],
): string {
  const normalizedDockerfile = stripLabelBlocks(dockerfileContent)
    .replace(/\r\n/g, '\n')
    .trim();
  const parts: string[] = [normalizedDockerfile];

  const sortedNames = Object.keys(copyFileContents).sort();
  for (const name of sortedNames) {
    parts.push(`${path.basename(name)}\0${sha256Hex(copyFileContents[name])}`);
  }
  parts.push(`MODULES\0${[...selectedModuleIds].sort().join(',')}`);

  return sha256Hex(parts.join('\n'));
}

export function fingerprintTag(fingerprint: string): string {
  return `devcontainer-cli/${fingerprint.slice(0, 12)}:latest`;
}

export function localImageExists(image: string): boolean {
  const r = dockerCapture(['image', 'inspect', image]);
  return r.status === 0;
}

export function recordProject(projectDir: string, config: DevcontainerConfig, customImage?: string): void {
  try {
    const now = new Date().toISOString();
    recordEntry({
      projectDir,
      workspace: config.workspace,
      mode: config.mode,
      image: customImage ?? config.image,
      variant: config.remote?.variant,
      fingerprint: config.fingerprint,
      createdAt: now,
      lastUpdated: now,
    });
  } catch {
    // Non-fatal
  }
}
